ledger/trie/trie.go

```diff
@@ -0,0 +1,321 @@
+				newLeaf := NewLeaf(nodeHeight(matched+1), path, payload)
+				t.store.Save(newLeaf.Hash(), payload)
+
+				var lChild, rChild Node
+				if bitutils.Bit(path[:], int(matched)) == 0 {
+					lChild = newLeaf
+					rChild = newExt
+				} else {
+					lChild = newExt
+					rChild = newLeaf
+				}
+
+				// Change children, path and skipped height of the original extension by recreating it.
+				*current = NewExtension(node.height, nodeHeight(matched), path, lChild, rChild)
+				return
+			}
+
+			// The path of the new leaf to insert matches with the skipped part of the extension's path, so we just
+			// treat it like a branch and skip through many depths at once.
+			if bitutils.Bit(path[:], int(nodeHeight(node.skip))) == 0 {
+				current = &node.lChild
+			} else {
+				current = &node.rChild
+			}
+			depth = nodeHeight(node.skip - 1)
+
+		case *Leaf:
+			if node.path == path {
+				// This path conflicts with a leaf, overwrite its hash using the new payload.
+				node.hash = ledger.ComputeCompactValue(hash.Hash(path), payload.Value, int(node.height))
+				t.store.Save(node.hash, payload)
+				return
+			}
+
+			// This leaf is currently at a height which conflicts with the new path that we want to insert.
+			// Therefore, we need to replace this leaf with a branch or extension that has two children, the
+			// new leaf and the previous one.
+
+			matched := commonBits(node.path, path)
+
+			// We need to fetch the payload here since the old leaf now resides at a new height and therefore its
+			// hash needs to be recomputed.
+			oldPayload, err := t.store.Retrieve(node.Hash())
+			if err != nil {
+				panic(err) // FIXME: Handle this gracefully.
+			}
+
+			oldLeaf := NewLeaf(nodeHeight(matched+1), node.path, oldPayload)
+			t.store.Save(oldLeaf.Hash(), oldPayload)
+
+			newLeaf := NewLeaf(nodeHeight(matched+1), path, payload)
+			t.store.Save(newLeaf.Hash(), payload)
+
+			// Compare first different bit between existing leaf and new leaf to know which one is which child for the
+			// newly created branch.
+			var lChild, rChild Node
+			if bitutils.Bit(path[:], int(matched)) == 0 {
+				lChild = newLeaf
+				rChild = oldLeaf
+			} else {
+				lChild = oldLeaf
+				rChild = newLeaf
+			}
+
+			// Create an extension node that skips up to the depth at which the old and
+			// new leaves diverge in path, or a branch if the extension would not end up skipping anything.
+			if depth == matched {
+				// Create a branch, since an extension would not skip anything here.
+				*current = NewBranch(nodeHeight(depth), lChild, rChild)
+			} else {
+				// Create an extension to skip over the common bits between both node paths.
+				*current = NewExtension(nodeHeight(depth), nodeHeight(matched), path, lChild, rChild)
+			}
+
+			return
+
+		case nil:
+			// There is no leaf here yet, create it.
+			*current = NewLeaf(nodeHeight(depth), path, payload)
+			t.store.Save((*current).Hash(), payload)
+			return
+		}
+	}
+}
+
```

Comment:
Some of the exported methods have comments, and some do not; not sure why (maybe you planned to add them later)